src/core/iter.rs

```diff
@@ -0,0 +1,98 @@
+// Rust Simplicity Library
+// Written in 2022 by
+//   Christian Lewe <clewe@blockstream.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Iterators for DAG-like structures
+
+use std::iter::Rev;
+use std::vec::IntoIter;
+
+/// A DAG-like structure with a root,
+/// left children and right children per node,
+/// without directed cycles.
+///
+/// We use this trait to implement general DAG operations, like node traversal.
+pub trait DagLike: Sized {
+    /// Node in the DAG
+    type Node: Copy;
+    /// Additional data for internal computations
+    type Extra;
```

Comment:
In 925b8cbdb0be693e1fa8d93005eb87d2a9c54846:

Would prefer this be called `Aux` rather than `Extra`